sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/WriteTables.java

```diff
@@ -106,17 +108,29 @@ public void processElement(ProcessContext c) throws Exception {
       dynamicDestinations.setSideInputAccessorFromProcessContext(c);
       DestinationT destination = c.element().getKey().getKey();
       TableSchema tableSchema;
-      String jsonSchema = jsonSchemas.get(destination);
-      if (jsonSchema != null) {
-        tableSchema = BigQueryHelpers.fromJsonString(jsonSchema, TableSchema.class);
+      if (firstPaneCreateDisposition == CreateDisposition.CREATE_NEVER) {
+        tableSchema = null;
+      } else if (jsonSchemas.containsKey(destination)) {
+        tableSchema =
+            BigQueryHelpers.fromJsonString(jsonSchemas.get(destination), TableSchema.class);
       } else {
         tableSchema = dynamicDestinations.getSchema(destination);
-        if (tableSchema != null) {
-          jsonSchemas.put(destination, BigQueryHelpers.toJsonString(tableSchema));
-        }
+        checkArgument(
+            tableSchema != null,
+            "DynamicDestinations.getSchema() may not return null, "
```

Comment:
Here also, we should probably mention that DynamicDestinations.getSchema() may return null if firstPaneCreateDisposition == CreateDisposition.CREATE_NEVER.